src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs

```diff
@@ -0,0 +1,137 @@
+        /// </summary>
+        /// <param name="amazonS3Client">Amazon S3 client</param>
+        /// <param name="logger">Microsoft.Extensions.Logging ILogger</param>
+        /// <param name="storageOptions">The S3 storage options</param>
+        public AWSS3StorageImageProvider(IAmazonS3 amazonS3Client, ILogger<AWSS3StorageImageProvider> logger, IOptions<AWSS3StorageImageProviderOptions> storageOptions)
+        {
+            Guard.NotNull(storageOptions, nameof(storageOptions));
+
+            this.amazonS3Client = amazonS3Client;
+            this.logger = logger;
+            this.storageOptions = storageOptions.Value;
+        }
+
+        /// <inheritdoc />
+        public Func<HttpContext, bool> Match
+        {
+            get => this.match ?? this.IsMatch;
+            set => this.match = value;
+        }
+
+        /// <inheritdoc />
+        public bool IsValidRequest(HttpContext context)
+        {
+            var displayUrl = context.Request.Path;
+            return Path.GetExtension(displayUrl).EndsWith(".jpg") || Path.GetExtension(displayUrl).EndsWith(".png");
+        }
+
+        /// <inheritdoc />
+        public async Task<IImageResolver> GetAsync(HttpContext context)
+        {
+            PathString displayUrl = context.Request.Path;
+            this.logger.LogDebug("Getting image for {ImageUri}", displayUrl);
+            string imageId = Path.GetFileNameWithoutExtension(displayUrl);
+            this.logger.LogDebug("Image id is {ImageId}", imageId);
+
+            string imagePath = $"{imageId}.png";
+
+            bool imageExists = await this.KeyExists(this.storageOptions.BucketName, imagePath);
+
+            if (!imageExists)
+            {
+                this.logger.LogDebug("No image found for {ImageId}", imageId);
+                return null;
+            }
+
+            this.logger.LogDebug("Found image {ImageId}", imageId);
+
+            return new AWSS3FileSystemResolver(this.amazonS3Client, this.storageOptions.BucketName, imagePath);
+        }
+
+        private bool IsMatch(HttpContext context)
+        {
+            string path = context.Request.Path.Value.TrimStart(SlashChars);
+            return path.StartsWith(this.storageOptions.BucketName, StringComparison.OrdinalIgnoreCase);
+        }
+
+        // ref https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Services/S3/Custom/_bcl/IO/S3FileInfo.cs#L118
+        private async Task<bool> KeyExists(string bucketName, string key, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            this.logger.LogDebug("Checking for the existence of key {Key} in bucket {BucketName}", key, bucketName);
+
+            try
+            {
+                var request = new GetObjectMetadataRequest
+                {
+                    BucketName = bucketName,
+                    Key = key
+                };
+
+                ((Amazon.Runtime.Internal.IAmazonWebServiceRequest)request)
```

Comment:
I don't this is correct. The internal namespace is a big red flag.

Reply:
This function came from their library. I will check but I think I can remove this section.